File: apps/expo-app/src/types/contracts.ts
export type EventEnvelope = {
  type: string;
  timestamp: string;
  source?: "sidecar_runtime" | "android_bridge_observation" | "ui_local";
  truth_scope?: "sidecar_health" | "sidecar_ready" | "native_runtime_host" | "task_route_surface" | "capability_state" | "ui_submit";
  payload: Record<string, any>;
};

export type RuntimeRecoveryResultState =
  | "restart_verified"
  | "restart_failed"
  | "restart_rate_limited"
  | "restart_needs_operator";

export type RuntimeRecoverySource = "sidecar_runtime" | "android_bridge_observation" | "ui_local";

export type RuntimeRecoveryTruthScope =
  | "sidecar_health"
  | "sidecar_ready"
  | "native_runtime_host"
  | "task_route_surface"
  | "ui_submit";

export type RuntimeRecoveryOutcome = {
  result: RuntimeRecoveryResultState;
  source: RuntimeRecoverySource;
  truth_scope: RuntimeRecoveryTruthScope[];
  verified: boolean;
  checked_at: string;
  note: string;
  blocking_reason?: string;
  attempts_used: number;
  next_allowed_at?: string;
};

export type ProviderDiagnosticPayload = {
  task_id?: string;
  attempt_id?: string;
  error_code?: string;
  code?: string;
  error_category?: string;
  category?: string;
  provider?: string;
  base_url_host?: string;
  endpoint_path?: string;
  http_status?: number;
  network_class?: string;
  provider_error_class?: string;
};

export type SafetyDecisionOutcome = "allow" | "block" | "clarify" | "confirm" | "safe_redirect";

export type SafetyDecisionReasonCode =
  | "none"
  | "harmful_request"
  | "prompt_injection_authority_spoof"
  | "external_content_attempted_command"
  | "missing_permission_or_capability"
  | "missing_tool_proof"
  | "missing_provider_access_route"
  | "missing_approval"
  | "unsafe_automation"
  | "destructive_action"
  | "privacy_surveillance_risk"
  | "credential_secret_risk"
  | "completion_evidence_missing"
  | "unknown_malformed_action"
  | "unbounded_consumption_risk";

export type SafetyDecisionGatePhase =
  | "pre_task"
  | "pre_context_injection"
  | "pre_model_action"
  | "pre_tool_action"
  | "pre_completion"
  | "pre_memory_write";

export type SafetyDecision = {
  outcome: SafetyDecisionOutcome;
  reason_code: SafetyDecisionReasonCode;
  gate_phase: SafetyDecisionGatePhase;
  action_family?: string;
  action_name?: string;
  evidence_required: boolean;
  source_trust_tier?: number;
  user_safe_message?: string;
  safe_summary?: string;
  created_at?: string;
};

export type IntakeAssessment = {
  primary_class?: string;
  secondary_flags?: string[];
  trust_tier?: number;
  validation_state?: string;
  chosen_closed_action?: string;
  memory_intent?: Record<string, unknown> | null;
  safety_decision?: SafetyDecision | null;
};

export type TaskSnapshot = {
  task_id: string;
  attempt_id: string;
  prompt: string;
  intent: string;
  room_id: string;
  anchor_id: string;
  status: string;
  started_at: string;
  updated_at: string;
  retry_count: number;
  max_retries: number;
  failure_type?: string;
  stuck_reason?: string;
  slot?: string;
  intake_assessment?: IntakeAssessment | null;
};

export type RuntimeState = {
  milo_state: string;
  current_room_id: string;
  current_anchor_id: string;
  last_meaningful_user_action_at: string;
  active_task?: TaskSnapshot | null;
  queued_task?: TaskSnapshot | null;
  capability_state?: Record<string, any> | null;
  runtime_id: string;
  pending_approval?: boolean | null;
  resume_checkpoint?: string | null;
};

export type CommandSubmitResponse = {
  handled?: boolean;
  kind?: "task" | "movement" | string;
  task_id?: string;
  attempt_id?: string;
  immediate_state?: TaskSnapshot | null;
  intake_gate?: SafetyDecision | Record<string, unknown> | null;
  plan?: Record<string, unknown> | null;
};

export type CueDescriptor = {
  voice_cue?: string;
  physical_cue?: string;
  description?: string;
};

export type RitualArtState = {
  status: string;
  title: string;
  chamber: string;
  description: string;
  cues: CueDescriptor;
};

export type RoomPresenceState = {
  room_id: string;
  anchor_id: string;
  milo_state: string;
};

export type MovementIntentState = {
  from_room?: string;
  from_anchor?: string;
  to_room?: string;
  to_anchor?: string;
  reason?: string;
  estimated_ms?: number;
};

export type ArtDegradationState = {
  reason?: string;
  surface?: string;
  message?: string;
};

export type ArtPresentationState = {
  task_state: string;
  room_presence: RoomPresenceState;
  movement_intent?: MovementIntentState | null;
  ritual_state?: RitualArtState | null;
  degradation_reason?: ArtDegradationState | null;
};

File: sidecar-go/shared/contracts/contracts.go
package contracts

type EventEnvelope struct {
	Type      string      `json:"type"`
	Timestamp string      `json:"timestamp"`
	Payload   interface{} `json:"payload"`
}

type TurnMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type RelayTurnRequest struct {
	TaskID           string        `json:"task_id"`
	Phase            string        `json:"phase"`
	Prompt           string        `json:"prompt"`
	SystemPrompt     string        `json:"system_prompt"`
	ConversationTail []TurnMessage `json:"conversation_tail"`
	ResponseStyle    string        `json:"response_style"`
}

type ExpectedCheck struct {
	CheckType     string `json:"check_type"`
	Key           string `json:"key,omitempty"`
	ExpectedValue string `json:"expected_value,omitempty"`
}

type UpdatedCheckpoint struct {
	Status    string `json:"status"`
	Blocker   string `json:"blocker,omitempty"`
	UpdatedAt string `json:"updated_at,omitempty"`
	ExpiresAt string `json:"expires_at,omitempty"`
}

type ExecutionResult struct {
	Status            string             `json:"status"`
	Verified          bool               `json:"verified"`
	ResultSummary     string             `json:"result_summary"`
	BlockingReason    string             `json:"blocking_reason,omitempty"`
	UpdatedCheckpoint *UpdatedCheckpoint `json:"updated_checkpoint,omitempty"`
}

type RelayTurnResponse struct {
	Intent             string           `json:"intent"`
	TargetRoom         string           `json:"target_room"`
	ThoughtText        string           `json:"thought_text"`
	Summary            string           `json:"summary"`
	ReportText         string           `json:"report_text"`
	CompletionStatus   string           `json:"completion_status"`
	ContinuationStatus string           `json:"continuation_status,omitempty"`
	NextBlocker        string           `json:"next_blocker,omitempty"`
	ActionType         string           `json:"action_type,omitempty"`
	ActionPayload      map[string]any   `json:"action_payload,omitempty"`
	ExpectedCheck      *ExpectedCheck   `json:"expected_check,omitempty"`
	ExecutionResult    *ExecutionResult `json:"execution_result,omitempty"`
	RequiresUserChoice bool             `json:"requires_user_choice"`
	Choices            []string         `json:"choices"`
}

type SessionStartRequest struct {
	DeviceUserID string `json:"device_user_id"`
	DeviceName   string `json:"device_name"`
	AppVersion   string `json:"app_version"`
}

type SessionStartResponse struct {
	DeviceUserID        string `json:"device_user_id"`
	SessionJWT          string `json:"session_jwt"`
	ExpiresAt           string `json:"expires_at"`
	Entitled            bool   `json:"entitled"`
	AccessMode          string `json:"access_mode"`
	AccessCodeOnly      bool   `json:"access_code_only"`
	TrialAllowed        bool   `json:"trial_allowed"`
	SubscriptionAllowed bool   `json:"subscription_allowed"`
	AccessCodeGrantDays int    `json:"access_code_grant_days"`
	VerifiedEmail       string `json:"verified_email,omitempty"`
	EmailVerified       bool   `json:"email_verified"`
	TwoFactorEnabled    bool   `json:"two_factor_enabled"`
	TwoFactorOK         bool   `json:"two_factor_ok"`
	WebsiteHandoffReady bool   `json:"website_handoff_ready"`
}

type CueDescriptor struct {
	VoiceCue    string `json:"voice_cue,omitempty"`
	PhysicalCue string `json:"physical_cue,omitempty"`
	Description string `json:"description,omitempty"`
}

type RitualArtState struct {
	Status      string        `json:"status"`
	Title       string        `json:"title"`
	Chamber     string        `json:"chamber"`
	Description string        `json:"description"`
	Cues        CueDescriptor `json:"cues"`
}

type RoomPresenceState struct {
	RoomID    string `json:"room_id"`
	AnchorID  string `json:"anchor_id"`
	MiloState string `json:"milo_state"`
}

type MovementIntentState struct {
	FromRoom    string `json:"from_room,omitempty"`
	FromAnchor  string `json:"from_anchor,omitempty"`
	ToRoom      string `json:"to_room,omitempty"`
	ToAnchor    string `json:"to_anchor,omitempty"`
	Reason      string `json:"reason,omitempty"`
	EstimatedMS int    `json:"estimated_ms,omitempty"`
}

type ArtDegradationState struct {
	Reason  string `json:"reason,omitempty"`
	Surface string `json:"surface,omitempty"`
	Message string `json:"message,omitempty"`
}

type ArtPresentationState struct {
	TaskState         string               `json:"task_state"`
	RoomPresence      RoomPresenceState    `json:"room_presence"`
	MovementIntent    *MovementIntentState `json:"movement_intent,omitempty"`
	RitualState       *RitualArtState      `json:"ritual_state,omitempty"`
	DegradationReason *ArtDegradationState `json:"degradation_reason,omitempty"`
}
